climakitae/uncertain_utils.py

```diff
@@ -0,0 +1,425 @@
+
+    # subsets the cmip6 and cal-adapt models in the future (ssp370) period
+    ssp_dsets = {key: val for key, val in dsets.items() if "ssp370" in key}
+    cal_ssp_dsets = {key: val for key, val in cal_dsets.items() if "ssp370" in key}
+
+    # drop member id, in order to ensure that merging is along the same data axis
+    hist_dsets = _drop_member_id(hist_dsets)
+    cal_hist_dsets = _drop_member_id(cal_hist_dsets)
+    ssp_dsets = _drop_member_id(ssp_dsets)
+    cal_ssp_dsets = _drop_member_id(cal_ssp_dsets)
+
+    # merge datasets together
+    all_hist_mdls = hist_dsets | cal_hist_dsets
+    all_ssp_mdls = ssp_dsets | cal_ssp_dsets
+
+    # concatenate historical data based on the model, and subset for California
+    hist_ds = xr.concat(list(all_hist_mdls.values()), dim="simulation").squeeze()
+    hist_ds = copt._cmip_clip(
+        hist_ds.sel(time=slice("1850", "2014"))
+    )  # time slice ensures the same historical timeframe
+    ssp_ds = xr.concat(list(all_ssp_mdls.values()), dim="simulation").squeeze()
+    ssp_ds = copt._cmip_clip(ssp_ds)
+
+    # concatenate all data together based on model
+    mdls_ds = xr.concat(
+        [hist_ds, ssp_ds],
+        dim="time",
+        coords="minimal",
+        compat="override",
+        join="inner",
+    )
+
+    return mdls_ds
+
+
+## -----------------------------------------------------------------------------
+## Useful individual analysis functions
+
+
+def cmip_annual(ds):
+    """Calculates the annual average temperature timeseries in degC from monthly data.
+
+    Note: as is, is specified for temperature in order to convert units. Can be
+    generalized.
+
+    Parameters
+    ----------
+    ds: xr.Dataset
+        Input data, default temperature unit is K
+
+    Returns
+    -------
+    xr.Dataset
+        Annual temperature timeseries in degC
+    """
+    ds_degC = ds - 273.15  # convert to degC
+    ds_degC = ds_degC.groupby("time.year").mean(dim=["time"])
+    return ds_degC
+
+
+def calc_anom(ds_yr, base_start, base_end):
+    """Calculates the difference relative to a historical baseline.
+
+    First calculates a baseline per simulation using input (base_start, base_end).
+    Then calculates the anomaly from baseline per simulation.
+
+    Parameters
+    ----------
+    ds_yr: xr.Dataset
+        must be the output from cmip_annual
+    base_start: int
+        start year of baseline to calculate
+    base_end: int
+        end year of the baseline to calculate
+
+    Returns
+    -------
+    xr.Dataset
+        Anomaly data calculated with input baseline start and end
+    """
+    mdl_baseline = ds_yr.sel(year=slice(base_start, base_end)).mean("year")
+    mdl_temp_anom = ds_yr - mdl_baseline
+    return mdl_temp_anom
+
+
+def cmip_mmm(ds):
+    """Calculate the CMIP6 multi-model mean by collapsing across simulations.
+
+    Parameters
+    ----------
+    ds: xr.Dataset
+        Input data, multiple simulations
+
```

Comment:
Just a sanity check: Are you sure that these functions all take/return Dataset objects and not DataArray objects?